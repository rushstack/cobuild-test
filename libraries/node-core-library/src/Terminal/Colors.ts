// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * @beta
 */
export interface IColorableSequence {
  text: string;
  isEol?: boolean;
  foregroundColor?: ColorValue;
  backgroundColor?: ColorValue;
  textAttributes?: TextAttribute[];
}

export const eolSequence: IColorableSequence = {
  isEol: true
} as IColorableSequence;

/**
 * Colors used with {@link IColorableSequence}.
 * @beta
 */
export enum ColorValue {
  Black,
  Red,
  Green,
  Yellow,
  Blue,
  Magenta,
  Cyan,
  White,
  Gray
}

/**
 * Text styles used with {@link IColorableSequence}.
 * @beta
 */
export enum TextAttribute {
  Bold,
  Dim,
  Underline,
  Blink,
  InvertColor,
  Hidden
}

export enum ConsoleColorCodes {
  BlackForeground = 30,
  RedForeground = 31,
  GreenForeground = 32,
  YellowForeground = 33,
  BlueForeground = 34,
  MagentaForeground = 35,
  CyanForeground = 36,
  WhiteForeground = 37,
  GrayForeground = 90,
  DefaultForeground = 39,

  BlackBackground = 40,
  RedBackground = 41,
  GreenBackground = 42,
  YellowBackground = 43,
  BlueBackground = 44,
  MagentaBackground = 45,
  CyanBackground = 46,
  WhiteBackground = 47,
  GrayBackground = 100,
  DefaultBackground = 49,

  Bold = 1,

  // On Linux, the "BoldOff" code instead causes the text to be double-underlined:
  // https://en.wikipedia.org/wiki/Talk:ANSI_escape_code#SGR_21%E2%80%94%60Bold_off%60_not_widely_supported
  // Use "NormalColorOrIntensity" instead
  // BoldOff = 21,

  Dim = 2,
  NormalColorOrIntensity = 22,
  Underline = 4,
  UnderlineOff = 24,
  Blink = 5,
  BlinkOff = 25,
  InvertColor = 7,
  InvertColorOff = 27,
  Hidden = 8,
  HiddenOff = 28
}

/**
 * The static functions on this class are used to produce colored text
 * for use with the node-core-library terminal.
 *
 * @example
 * terminal.writeLine(Colors.green('Green Text!'), ' ', Colors.blue('Blue Text!'));
 *
 * @beta
 */
export class Colors {
  public static black(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Black
    };
  }

  public static red(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Red
    };
  }

  public static green(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Green
    };
  }

  public static yellow(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Yellow
    };
  }

  public static blue(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Blue
    };
  }

  public static magenta(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Magenta
    };
  }

  public static cyan(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Cyan
    };
  }

  public static white(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.White
    };
  }

  public static gray(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Gray
    };
  }

  public static blackBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Black
    };
  }

  public static redBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Red
    };
  }

  public static greenBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Green
    };
  }

  public static yellowBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Yellow
    };
  }

  public static blueBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Blue
    };
  }

  public static magentaBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Magenta
    };
  }

  public static cyanBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Cyan
    };
  }

  public static whiteBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.White
    };
  }

  public static grayBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Gray
    };
  }

  public static bold(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Bold);
  }

  public static dim(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Dim);
  }

  public static underline(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Underline);
  }

  public static blink(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Blink);
  }

  public static invertColor(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.InvertColor);
  }

  public static hidden(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Hidden);
  }

  /**
   * If called with a string, returns the string wrapped in a {@link IColorableSequence}.
   * If called with a {@link IColorableSequence}, returns the {@link IColorableSequence}.
   *
   * @internal
   */
  public static _normalizeStringOrColorableSequence(value: string | IColorableSequence): IColorableSequence {
    if (typeof value === 'string') {
      return {
        text: value
      };
    } else {
      return value;
    }
  }

  private static _applyTextAttribute(
    text: string | IColorableSequence,
    attribute: TextAttribute
  ): IColorableSequence {
    const sequence: IColorableSequence = Colors._normalizeStringOrColorableSequence(text);
    if (!sequence.textAttributes) {
      sequence.textAttributes = [];
    }

    sequence.textAttributes.push(attribute);
    return sequence;
  }
}
