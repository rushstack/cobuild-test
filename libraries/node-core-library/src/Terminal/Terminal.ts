// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ITerminalProvider, TerminalProviderSeverity } from './ITerminalProvider';
import {
  IColorableSequence,
  ColorValue,
  Colors,
  eolSequence,
  TextAttribute,
  ConsoleColorCodes
} from './Colors';
import { ITerminal } from './ITerminal';

/**
 * This class facilitates writing to a console.
 *
 * @beta
 */
export class Terminal implements ITerminal {
  private _providers: Set<ITerminalProvider>;

  public constructor(provider: ITerminalProvider) {
    this._providers = new Set<ITerminalProvider>();
    this._providers.add(provider);
  }

  /**
   * {@inheritdoc ITerminal.registerProvider}
   */
  public registerProvider(provider: ITerminalProvider): void {
    this._providers.add(provider);
  }

  /**
   * {@inheritdoc ITerminal.unregisterProvider}
   */
  public unregisterProvider(provider: ITerminalProvider): void {
    if (this._providers.has(provider)) {
      this._providers.delete(provider);
    }
  }

  /**
   * {@inheritdoc ITerminal.write}
   */
  public write(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.log);
  }

  /**
   * {@inheritdoc ITerminal.writeLine}
   */
  public writeLine(...messageParts: (string | IColorableSequence)[]): void {
    this.write(...messageParts, eolSequence);
  }

  /**
   * {@inheritdoc ITerminal.writeWarning}
   */
  public writeWarning(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      messageParts.map(
        (part): IColorableSequence => ({
          ...Colors._normalizeStringOrColorableSequence(part),
          foregroundColor: ColorValue.Yellow
        })
      ),
      TerminalProviderSeverity.warning
    );
  }

  /**
   * {@inheritdoc ITerminal.writeWarningLine}
   */
  public writeWarningLine(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      [
        ...messageParts.map(
          (part): IColorableSequence => ({
            ...Colors._normalizeStringOrColorableSequence(part),
            foregroundColor: ColorValue.Yellow
          })
        ),
        eolSequence
      ],
      TerminalProviderSeverity.warning
    );
  }

  /**
   * {@inheritdoc ITerminal.writeError}
   */
  public writeError(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      messageParts.map(
        (part): IColorableSequence => ({
          ...Colors._normalizeStringOrColorableSequence(part),
          foregroundColor: ColorValue.Red
        })
      ),
      TerminalProviderSeverity.error
    );
  }

  /**
   * {@inheritdoc ITerminal.writeErrorLine}
   */
  public writeErrorLine(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      [
        ...messageParts.map(
          (part): IColorableSequence => ({
            ...Colors._normalizeStringOrColorableSequence(part),
            foregroundColor: ColorValue.Red
          })
        ),
        eolSequence
      ],
      TerminalProviderSeverity.error
    );
  }

  /**
   * {@inheritdoc ITerminal.writeVerbose}
   */
  public writeVerbose(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.verbose);
  }

  /**
   * {@inheritdoc ITerminal.writeVerboseLine}
   */
  public writeVerboseLine(...messageParts: (string | IColorableSequence)[]): void {
    this.writeVerbose(...messageParts, eolSequence);
  }

  /**
   * {@inheritdoc ITerminal.writeDebug}
   */
  public writeDebug(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.debug);
  }

  /**
   * {@inheritdoc ITerminal.writeDebugLine}
   */
  public writeDebugLine(...messageParts: (string | IColorableSequence)[]): void {
    this.writeDebug(...messageParts, eolSequence);
  }

  private _writeSegmentsToProviders(
    segments: (string | IColorableSequence)[],
    severity: TerminalProviderSeverity
  ): void {
    const withColorText: { [eolChar: string]: string } = {};
    const withoutColorText: { [eolChar: string]: string } = {};
    let withColorLines: string[] | undefined;
    let withoutColorLines: string[] | undefined;

    this._providers.forEach((provider) => {
      const eol: string = provider.eolCharacter;
      let textToWrite: string;
      if (provider.supportsColor) {
        if (!withColorLines) {
          withColorLines = this._serializeFormattableTextSegments(segments, true);
        }

        if (!withColorText[eol]) {
          withColorText[eol] = withColorLines.join(eol);
        }

        textToWrite = withColorText[eol];
      } else {
        if (!withoutColorLines) {
          withoutColorLines = this._serializeFormattableTextSegments(segments, false);
        }

        if (!withoutColorText[eol]) {
          withoutColorText[eol] = withoutColorLines.join(eol);
        }

        textToWrite = withoutColorText[eol];
      }

      provider.write(textToWrite, severity);
    });
  }

  private _serializeFormattableTextSegments(
    segments: (string | IColorableSequence)[],
    withColor: boolean
  ): string[] {
    const lines: string[] = [];
    let segmentsToJoin: string[] = [];
    let lastSegmentWasEol: boolean = false;
    for (let i: number = 0; i < segments.length; i++) {
      const segment: IColorableSequence = Colors._normalizeStringOrColorableSequence(segments[i]);
      lastSegmentWasEol = !!segment.isEol;
      if (lastSegmentWasEol) {
        lines.push(segmentsToJoin.join(''));
        segmentsToJoin = [];
      } else {
        if (withColor) {
          const startColorCodes: number[] = [];
          const endColorCodes: number[] = [];
          switch (segment.foregroundColor) {
            case ColorValue.Black: {
              startColorCodes.push(ConsoleColorCodes.BlackForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Red: {
              startColorCodes.push(ConsoleColorCodes.RedForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Green: {
              startColorCodes.push(ConsoleColorCodes.GreenForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Yellow: {
              startColorCodes.push(ConsoleColorCodes.YellowForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Blue: {
              startColorCodes.push(ConsoleColorCodes.BlueForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Magenta: {
              startColorCodes.push(ConsoleColorCodes.MagentaForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Cyan: {
              startColorCodes.push(ConsoleColorCodes.CyanForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.White: {
              startColorCodes.push(ConsoleColorCodes.WhiteForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Gray: {
              startColorCodes.push(ConsoleColorCodes.GrayForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }
          }

          switch (segment.backgroundColor) {
            case ColorValue.Black: {
              startColorCodes.push(ConsoleColorCodes.BlackBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Red: {
              startColorCodes.push(ConsoleColorCodes.RedBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Green: {
              startColorCodes.push(ConsoleColorCodes.GreenBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Yellow: {
              startColorCodes.push(ConsoleColorCodes.YellowBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Blue: {
              startColorCodes.push(ConsoleColorCodes.BlueBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Magenta: {
              startColorCodes.push(ConsoleColorCodes.MagentaBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Cyan: {
              startColorCodes.push(ConsoleColorCodes.CyanBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.White: {
              startColorCodes.push(ConsoleColorCodes.WhiteBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Gray: {
              startColorCodes.push(ConsoleColorCodes.GrayBackground);
              endColorCodes.push(49);
              break;
            }
          }

          if (segment.textAttributes) {
            for (const textAttribute of segment.textAttributes) {
              switch (textAttribute) {
                case TextAttribute.Bold: {
                  startColorCodes.push(ConsoleColorCodes.Bold);
                  endColorCodes.push(ConsoleColorCodes.NormalColorOrIntensity);
                  break;
                }

                case TextAttribute.Dim: {
                  startColorCodes.push(ConsoleColorCodes.Dim);
                  endColorCodes.push(ConsoleColorCodes.NormalColorOrIntensity);
                  break;
                }

                case TextAttribute.Underline: {
                  startColorCodes.push(ConsoleColorCodes.Underline);
                  endColorCodes.push(ConsoleColorCodes.UnderlineOff);
                  break;
                }

                case TextAttribute.Blink: {
                  startColorCodes.push(ConsoleColorCodes.Blink);
                  endColorCodes.push(ConsoleColorCodes.BlinkOff);
                  break;
                }

                case TextAttribute.InvertColor: {
                  startColorCodes.push(ConsoleColorCodes.InvertColor);
                  endColorCodes.push(ConsoleColorCodes.InvertColorOff);
                  break;
                }

                case TextAttribute.Hidden: {
                  startColorCodes.push(ConsoleColorCodes.Hidden);
                  endColorCodes.push(ConsoleColorCodes.HiddenOff);
                  break;
                }
              }
            }
          }

          for (let j: number = 0; j < startColorCodes.length; j++) {
            const code: number = startColorCodes[j];
            segmentsToJoin.push(...['\u001b[', code.toString(), 'm']);
          }

          segmentsToJoin.push(segment.text);

          for (let j: number = endColorCodes.length - 1; j >= 0; j--) {
            const code: number = endColorCodes[j];
            segmentsToJoin.push(...['\u001b[', code.toString(), 'm']);
          }
        } else {
          segmentsToJoin.push(segment.text);
        }
      }
    }

    if (segmentsToJoin.length > 0) {
      lines.push(segmentsToJoin.join(''));
    }

    if (lastSegmentWasEol) {
      lines.push('');
    }

    return lines;
  }
}
