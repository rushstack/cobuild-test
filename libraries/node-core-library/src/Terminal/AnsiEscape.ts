// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ConsoleColorCodes } from './Colors';

/**
 * Options for {@link AnsiEscape.formatForTests}.
 * @public
 */
export interface IAnsiEscapeConvertForTestsOptions {
  /**
   * If true then `\n` will be replaced by `[n]`, and `\r` will be replaced by `[r]`.
   */
  encodeNewlines?: boolean;
}

/**
 * Operations for working with text strings that contain
 * {@link https://en.wikipedia.org/wiki/ANSI_escape_code | ANSI escape codes}.
 * The most commonly used escape codes set the foreground/background color for console output.
 * @public
 */
export class AnsiEscape {
  // For now, we only care about the Control Sequence Introducer (CSI) commands which always start with "[".
  // eslint-disable-next-line no-control-regex
  private static readonly _csiRegExp: RegExp = /\x1b\[([\x30-\x3f]*[\x20-\x2f]*[\x40-\x7e])/gu;

  // Text coloring is performed using Select Graphic Rendition (SGR) codes, which come after the
  // CSI introducer "ESC [".  The SGR sequence is a number followed by "m".
  private static readonly _sgrRegExp: RegExp = /([0-9]+)m/u;

  private static readonly _backslashNRegExp: RegExp = /\n/g;
  private static readonly _backslashRRegExp: RegExp = /\r/g;

  /**
   * Returns the input text with all ANSI escape codes removed.  For example, this is useful when saving
   * colorized console output to a log file.
   */
  public static removeCodes(text: string): string {
    // eslint-disable-next-line no-control-regex
    return text.replace(AnsiEscape._csiRegExp, '');
  }

  /**
   * Replaces ANSI escape codes with human-readable tokens.  This is useful for unit tests
   * that compare text strings in test assertions or snapshot files.
   */
  public static formatForTests(text: string, options?: IAnsiEscapeConvertForTestsOptions): string {
    if (!options) {
      options = {};
    }

    let result: string = text.replace(AnsiEscape._csiRegExp, (capture: string, csiCode: string) => {
      // If it is an SGR code, then try to show a friendly token
      const match: RegExpMatchArray | null = csiCode.match(AnsiEscape._sgrRegExp);
      if (match) {
        const sgrParameter: number = parseInt(match[1]);
        const sgrParameterName: string | undefined = AnsiEscape._tryGetSgrFriendlyName(sgrParameter);
        if (sgrParameterName) {
          // Example: "[black-bg]"
          return `[${sgrParameterName}]`;
        }
      }

      // Otherwise show the raw code, but without the "[" from the CSI prefix
      // Example: "[31m]"
      return `[${csiCode}]`;
    });

    if (options.encodeNewlines) {
      result = result
        .replace(AnsiEscape._backslashNRegExp, '[n]')
        .replace(AnsiEscape._backslashRRegExp, `[r]`);
    }
    return result;
  }

  // Returns a human-readable token representing an SGR parameter, or undefined for parameter that is not well-known.
  // The SGR parameter numbers are documented in this table:
  // https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_parameters
  private static _tryGetSgrFriendlyName(sgiParameter: number): string | undefined {
    switch (sgiParameter) {
      case ConsoleColorCodes.BlackForeground:
        return 'black';
      case ConsoleColorCodes.RedForeground:
        return 'red';
      case ConsoleColorCodes.GreenForeground:
        return 'green';
      case ConsoleColorCodes.YellowForeground:
        return 'yellow';
      case ConsoleColorCodes.BlueForeground:
        return 'blue';
      case ConsoleColorCodes.MagentaForeground:
        return 'magenta';
      case ConsoleColorCodes.CyanForeground:
        return 'cyan';
      case ConsoleColorCodes.WhiteForeground:
        return 'white';
      case ConsoleColorCodes.GrayForeground:
        return 'gray';
      case ConsoleColorCodes.DefaultForeground:
        return 'default';

      case ConsoleColorCodes.BlackBackground:
        return 'black-bg';
      case ConsoleColorCodes.RedBackground:
        return 'red-bg';
      case ConsoleColorCodes.GreenBackground:
        return 'green-bg';
      case ConsoleColorCodes.YellowBackground:
        return 'yellow-bg';
      case ConsoleColorCodes.BlueBackground:
        return 'blue-bg';
      case ConsoleColorCodes.MagentaBackground:
        return 'magenta-bg';
      case ConsoleColorCodes.CyanBackground:
        return 'cyan-bg';
      case ConsoleColorCodes.WhiteBackground:
        return 'white-bg';
      case ConsoleColorCodes.GrayBackground:
        return 'gray-bg';
      case ConsoleColorCodes.DefaultBackground:
        return 'default-bg';

      case ConsoleColorCodes.Bold:
        return 'bold';
      case ConsoleColorCodes.Dim:
        return 'dim';
      case ConsoleColorCodes.NormalColorOrIntensity:
        return 'normal';
      case ConsoleColorCodes.Underline:
        return 'underline';
      case ConsoleColorCodes.UnderlineOff:
        return 'underline-off';
      case ConsoleColorCodes.Blink:
        return 'blink';
      case ConsoleColorCodes.BlinkOff:
        return 'blink-off';
      case ConsoleColorCodes.InvertColor:
        return 'invert';
      case ConsoleColorCodes.InvertColorOff:
        return 'invert-off';
      case ConsoleColorCodes.Hidden:
        return 'hidden';
      case ConsoleColorCodes.HiddenOff:
        return 'hidden-off';
      default:
        return undefined;
    }
  }
}
