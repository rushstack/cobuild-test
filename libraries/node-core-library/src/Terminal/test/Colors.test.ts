// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Terminal } from '../Terminal';
import { StringBufferTerminalProvider } from '../StringBufferTerminalProvider';
import { createColorGrid } from './createColorGrid';
import { AnsiEscape } from '../AnsiEscape';
import { Colors } from '../Colors';

describe(Colors.name, () => {
  let terminal: Terminal;
  let provider: StringBufferTerminalProvider;

  beforeEach(() => {
    provider = new StringBufferTerminalProvider(true);
    terminal = new Terminal(provider);
  });

  test('writes color grid correctly', () => {
    for (const line of createColorGrid()) {
      terminal.writeLine(...line);
    }

    expect(provider.getOutput()).toMatchSnapshot();
  });

  test('correctly normalizes color codes for tests', () => {
    for (const line of createColorGrid()) {
      terminal.writeLine(...line);
    }

    expect(
      AnsiEscape.formatForTests(provider.getOutput({ normalizeSpecialCharacters: false }))
    ).toMatchSnapshot();
  });
});
