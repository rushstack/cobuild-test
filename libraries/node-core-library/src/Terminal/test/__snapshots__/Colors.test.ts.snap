// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colors correctly normalizes color codes for tests 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n][gray-bg]X[default-bg][black][gray-bg]X[default-bg][default][white][gray-bg]X[default-bg][default][gray][gray-bg]X[default-bg][default][magenta][gray-bg]X[default-bg][default][red][gray-bg]X[default-bg][default][yellow][gray-bg]X[default-bg][default][green][gray-bg]X[default-bg][default][cyan][gray-bg]X[default-bg][default][blue][gray-bg]X[default-bg][default][n][magenta-bg]X[default-bg][black][magenta-bg]X[default-bg][default][white][magenta-bg]X[default-bg][default][gray][magenta-bg]X[default-bg][default][magenta][magenta-bg]X[default-bg][default][red][magenta-bg]X[default-bg][default][yellow][magenta-bg]X[default-bg][default][green][magenta-bg]X[default-bg][default][cyan][magenta-bg]X[default-bg][default][blue][magenta-bg]X[default-bg][default][n][red-bg]X[default-bg][black][red-bg]X[default-bg][default][white][red-bg]X[default-bg][default][gray][red-bg]X[default-bg][default][magenta][red-bg]X[default-bg][default][red][red-bg]X[default-bg][default][yellow][red-bg]X[default-bg][default][green][red-bg]X[default-bg][default][cyan][red-bg]X[default-bg][default][blue][red-bg]X[default-bg][default][n][yellow-bg]X[default-bg][black][yellow-bg]X[default-bg][default][white][yellow-bg]X[default-bg][default][gray][yellow-bg]X[default-bg][default][magenta][yellow-bg]X[default-bg][default][red][yellow-bg]X[default-bg][default][yellow][yellow-bg]X[default-bg][default][green][yellow-bg]X[default-bg][default][cyan][yellow-bg]X[default-bg][default][blue][yellow-bg]X[default-bg][default][n][green-bg]X[default-bg][black][green-bg]X[default-bg][default][white][green-bg]X[default-bg][default][gray][green-bg]X[default-bg][default][magenta][green-bg]X[default-bg][default][red][green-bg]X[default-bg][default][yellow][green-bg]X[default-bg][default][green][green-bg]X[default-bg][default][cyan][green-bg]X[default-bg][default][blue][green-bg]X[default-bg][default][n][cyan-bg]X[default-bg][black][cyan-bg]X[default-bg][default][white][cyan-bg]X[default-bg][default][gray][cyan-bg]X[default-bg][default][magenta][cyan-bg]X[default-bg][default][red][cyan-bg]X[default-bg][default][yellow][cyan-bg]X[default-bg][default][green][cyan-bg]X[default-bg][default][cyan][cyan-bg]X[default-bg][default][blue][cyan-bg]X[default-bg][default][n][blue-bg]X[default-bg][black][blue-bg]X[default-bg][default][white][blue-bg]X[default-bg][default][gray][blue-bg]X[default-bg][default][magenta][blue-bg]X[default-bg][default][red][blue-bg]X[default-bg][default][yellow][blue-bg]X[default-bg][default][green][blue-bg]X[default-bg][default][cyan][blue-bg]X[default-bg][default][blue][blue-bg]X[default-bg][default][n]"`;

exports[`Colors writes color grid correctly 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n][gray-bg]X[default-bg][black][gray-bg]X[default-bg][default][white][gray-bg]X[default-bg][default][gray][gray-bg]X[default-bg][default][magenta][gray-bg]X[default-bg][default][red][gray-bg]X[default-bg][default][yellow][gray-bg]X[default-bg][default][green][gray-bg]X[default-bg][default][cyan][gray-bg]X[default-bg][default][blue][gray-bg]X[default-bg][default][n][magenta-bg]X[default-bg][black][magenta-bg]X[default-bg][default][white][magenta-bg]X[default-bg][default][gray][magenta-bg]X[default-bg][default][magenta][magenta-bg]X[default-bg][default][red][magenta-bg]X[default-bg][default][yellow][magenta-bg]X[default-bg][default][green][magenta-bg]X[default-bg][default][cyan][magenta-bg]X[default-bg][default][blue][magenta-bg]X[default-bg][default][n][red-bg]X[default-bg][black][red-bg]X[default-bg][default][white][red-bg]X[default-bg][default][gray][red-bg]X[default-bg][default][magenta][red-bg]X[default-bg][default][red][red-bg]X[default-bg][default][yellow][red-bg]X[default-bg][default][green][red-bg]X[default-bg][default][cyan][red-bg]X[default-bg][default][blue][red-bg]X[default-bg][default][n][yellow-bg]X[default-bg][black][yellow-bg]X[default-bg][default][white][yellow-bg]X[default-bg][default][gray][yellow-bg]X[default-bg][default][magenta][yellow-bg]X[default-bg][default][red][yellow-bg]X[default-bg][default][yellow][yellow-bg]X[default-bg][default][green][yellow-bg]X[default-bg][default][cyan][yellow-bg]X[default-bg][default][blue][yellow-bg]X[default-bg][default][n][green-bg]X[default-bg][black][green-bg]X[default-bg][default][white][green-bg]X[default-bg][default][gray][green-bg]X[default-bg][default][magenta][green-bg]X[default-bg][default][red][green-bg]X[default-bg][default][yellow][green-bg]X[default-bg][default][green][green-bg]X[default-bg][default][cyan][green-bg]X[default-bg][default][blue][green-bg]X[default-bg][default][n][cyan-bg]X[default-bg][black][cyan-bg]X[default-bg][default][white][cyan-bg]X[default-bg][default][gray][cyan-bg]X[default-bg][default][magenta][cyan-bg]X[default-bg][default][red][cyan-bg]X[default-bg][default][yellow][cyan-bg]X[default-bg][default][green][cyan-bg]X[default-bg][default][cyan][cyan-bg]X[default-bg][default][blue][cyan-bg]X[default-bg][default][n][blue-bg]X[default-bg][black][blue-bg]X[default-bg][default][white][blue-bg]X[default-bg][default][gray][blue-bg]X[default-bg][default][magenta][blue-bg]X[default-bg][default][red][blue-bg]X[default-bg][default][yellow][blue-bg]X[default-bg][default][green][blue-bg]X[default-bg][default][cyan][blue-bg]X[default-bg][default][blue][blue-bg]X[default-bg][default][n]"`;
